x-pack/legacy/plugins/ml/server/routes/modules.ts

```diff
@@ -0,0 +1,240 @@
+
+        return response.ok({ body: results });
+      } catch (e) {
+        return response.customError(wrapError(e));
+      }
+    })
+  );
+
+  /**
+   * @apiGroup DataRecognizer
+   *
+   * @api {get} /api/ml/modules/get_module/:moduleId Get module
+   * @apiName GetModule
+   * @apiDescription Returns module by id.
+   *
+   * @apiParam {String} [moduleId] Module id
+   */
+  router.get(
+    {
+      path: '/api/ml/modules/get_module/{moduleId?}',
+      validate: {
+        params: schema.object({
+          moduleId: schema.maybe(schema.string()),
+        }),
+      },
+    },
+    licensePreRoutingFactory(xpackMainPlugin, async (context, request, response) => {
+      try {
+        let { moduleId } = request.params;
+        if (moduleId === '') {
+          // if the endpoint is called with a trailing /
+          // the moduleId will be an empty string.
+          moduleId = undefined;
+        }
+        const results = await getModule(context, moduleId);
+
+        return response.ok({ body: results });
+      } catch (e) {
+        return response.customError(wrapError(e));
+      }
+    })
+  );
+
+  /**
+   * @apiGroup DataRecognizer
+   *
+   * @api {post} /api/ml/modules/setup/:moduleId Setup module
+   * @apiName SetupModule
+   * @apiDescription Created module items.
+   *
+   * @apiParam {String} moduleId Module id
+   */
+  router.post(
+    {
+      path: '/api/ml/modules/setup/{moduleId}',
+      validate: {
+        params: schema.object({
+          moduleId: schema.string(),
+        }),
+        body: schema.object({
+          prefix: schema.string(),
+          groups: schema.maybe(schema.arrayOf(schema.string())),
+          indexPatternName: schema.string(),
+          query: schema.maybe(schema.any()),
+          useDedicatedIndex: schema.maybe(schema.boolean()),
+          startDatafeed: schema.maybe(schema.boolean()),
+          start: schema.maybe(schema.number()),
+          end: schema.maybe(schema.number()),
+          jobOverrides: schema.maybe(schema.any()),
+          datafeedOverrides: schema.maybe(schema.any()),
+        }),
+      },
+    },
+    licensePreRoutingFactory(xpackMainPlugin, async (context, request, response) => {
+      try {
+        const { moduleId } = request.params;
+
+        const {
+          prefix,
+          groups,
+          indexPatternName,
+          query,
+          useDedicatedIndex,
+          startDatafeed,
+          start,
+          end,
+          jobOverrides,
+          datafeedOverrides,
+        } = request.body;
+
+        const result = await saveModuleItems(
+          context,
+          moduleId,
+          prefix,
+          groups,
+          indexPatternName,
+          query,
+          useDedicatedIndex,
+          startDatafeed,
+          start,
+          end,
+          jobOverrides,
+          datafeedOverrides,
+          getSavedObjectsClient(elasticsearchPlugin, savedObjects!)
```

Comment:
i'm unsure about using `getSavedObjectsClient` here as it uses the elevated internal user to create the saved object repository which could be a security risk.
The original version used the `getSavedObjectsClient` from the `request` which I imagine had the current user's level of privileges.
I'm unsure about this and I think is worth investigating.
Also, I didn't realise that the `savedObjects` being passed to the router are of the type `SavedObjectsLegacyService`.
I wonder if it's possible to use the non-legacy version of the saved objects client in our plugin right now, before we move it to the new platform proper.

Reply:
Agreed 🤔 I think we should be to use NP's `context.core.savedObjects.client` which is meant to replace `request.getSavedObjectsClient` (https://github.com/elastic/kibana/blob/master/src/core/MIGRATION.md#server-side)

Since `context` is scoped like the previously used `request` this should give us the same behavior, I'd think.

Reply:
Fixed in https://github.com/elastic/kibana/pull/57206/commits/63ea75b825f7659b09b62f62ab420f33651637ed